File: mobile/csm_app/lib/features/admin/projects/data/models/milestones/milestone.dart
class Milestone {
  const Milestone({
    required this.id,
    required this.companyId,
    required this.projectId,
    this.projectPhaseId,
    required this.name,
    this.description,
    required this.plannedDate,
    this.completedDate,
    required this.isCompleted,
    required this.createdAtUtc,
  });

  final String id;
  final String companyId;
  final String projectId;
  final String? projectPhaseId;
  final String name;
  final String? description;
  final DateTime plannedDate;
  final DateTime? completedDate;
  final bool isCompleted;
  final DateTime createdAtUtc;

  factory Milestone.fromJson(Map<String, dynamic> json) {
    return Milestone(
      id: json['id'] as String,
      companyId: json['companyId'] as String,
      projectId: json['projectId'] as String,
      projectPhaseId: json['projectPhaseId'] as String?,
      name: json['name'] as String,
      description: json['description'] as String?,
      plannedDate: DateTime.parse(
        json['plannedDate'] as String,
      ),
      completedDate: _parseDate(json['completedDate']),
      isCompleted: json['isCompleted'] as bool,
      createdAtUtc: DateTime.parse(
        json['createdAtUtc'] as String,
      ),
    );
  }

  static DateTime? _parseDate(dynamic value) {
    if (value == null) {
      return null;
    }

    return DateTime.tryParse(value.toString());
  }
}

File: src/CSM.Domain/Entities/Projects/Milestone.cs
using CSM.Domain.Common;

namespace CSM.Domain.Entities.Projects;

public class Milestone : TenantEntity
{
    public Guid ProjectId { get; set; }

    public Guid? ProjectPhaseId { get; set; }

    public string Name { get; set; } = string.Empty;

    public string? Description { get; set; }

    public string MilestoneNumber { get; set; } = string.Empty;

    public DateOnly PlannedDate { get; set; }

    public DateOnly? CompletedDate { get; set; }

    public bool IsCompleted { get; set; }

    public Project Project { get; set; } = null!;

    public ProjectPhase? ProjectPhase { get; set; }
}